src/components/Table/utils/TableState.ts

```diff
@@ -218,25 +218,61 @@ export class TableState {
   setColumns(columns: GridColumnWithId<any>[], visibleColumnsStorageKey: string | undefined): void {
     const isInitial = !this.columns || this.columns.length === 0;
     if (columns !== this.columns) {
-      this.columns = columns;
       this.visibleColumnsStorageKey = visibleColumnsStorageKey ?? camelCase(columns.map((c) => c.id).join());
       this.visibleColumns.replace(readOrSetLocalVisibleColumnState(columns, this.visibleColumnsStorageKey));
-      const expandedColumnIds = columns.filter((c) => c.initExpanded).map((c) => c.id);
-      if (isInitial) this.expandedColumns.replace(expandedColumnIds);
+      // list of local storage columns
+      const localStorageColumns = this.persistCollapse ? readExpandedColumnsStorage(this.persistCollapse) : [];
+      // expandedColumnIds initialized as localStorageColumns
+      const existingExpandedColumnIds = isInitial ? localStorageColumns : this.expandedColumnIds;
+
+      // list of all existing columns
+      const existingColumnIds = this.columns.map((c) => c.id);
+
+      // filter existing columns that are initExpanded to see which are new
+      const newExpandedColumnsIds = columns
+        .filter((c) => !existingColumnIds.includes(c.id) && c.initExpanded)
+        .map((c) => c.id);
+
+      const isDifferent = !newExpandedColumnsIds.every((c) => existingExpandedColumnIds.includes(c));
+      if (isDifferent) {
+        existingExpandedColumnIds.push(...newExpandedColumnsIds);
+      }
+
+      const newExpanded = columns.filter((c) => existingExpandedColumnIds.includes(c.id));
+      // if we have existing expanded columns ids, then do a Promise.all() and execute column.expandColumns
+      Promise.all(
+        newExpanded.map(async (c) => {
+          await this.loadExpandedColumns(c);
+        }),
+      ).then(() => {
+        const newIds = newExpanded.map((c) => c.id);
+        // if there is a difference between list of current expanded columns vs list we just created, then replace
+
+        if (isInitial || isDifferent) {
+          this.expandedColumns.replace(newIds);
+        }
+        //  update our persistCollapse if set
+        if (this.persistCollapse) {
+          sessionStorage.setItem(`expandedColumn_${this.persistCollapse}`, JSON.stringify(newIds));
```

Comment:
**suggestion:** We generate this string four times in this file, so I think it'd be worth making some helpers here. Maybe something along the lines of:
```
sessionStorage.setItem(getColumnStorageKey(this.persistCollapse), JSON.stringify(newIds));

...

function getColumnStorageKey(storageKey: string): string {
  return `expandedColumn_${storageKey}`;
}
```

Reply:
got it